pkg/frontend/querymiddleware/reject.go

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+type rejectMiddleware struct {
+	next   MetricsQueryHandler
+	limits Limits
+	logger log.Logger
+}
+
+// newRejectMiddleware creates a middleware that blocks queries that contain PromQL experimental functions
+// that are not enabled for the active tenant, allowing us to enable them only for selected tenants.
+func newRejectMiddleware(limits Limits, logger log.Logger) MetricsQueryMiddleware {
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &rejectMiddleware{
+			next:   next,
+			limits: limits,
+			logger: logger,
+		}
+	})
+}
+
+func (rm *rejectMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	// log := spanlogger.FromContext(ctx, rm.logger)
+
+	if !parser.EnableExperimentalFunctions {
+		// If experimental functions are disabled globally, we don't need to check for tenant-specific
+		// settings, and can skip this middleware and leave the checking to the rest of the flow.
+		return rm.next.Do(ctx, req)
+	}
+
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	experimentalFunctionsEnabled := validation.AllTrueBooleansPerTenant(tenantIDs, rm.limits.PromQLExperimentalFunctionsEnabled)
+
+	if experimentalFunctionsEnabled {
+		// If experimental functions are enabled for this tenant, we don't need to check the query
+		// for those functions and can skip this middleware.
+		return rm.next.Do(ctx, req)
+	}
+
+	expr, err := parser.ParseExpr(req.GetQuery())
```

Comment:
I've exposed the `parser.Expr` member in this PR, but I wonder if it's safe to do so. Any idea what's the reason it wasn't done before? We seem to run `parser.ParseExpr` a fair bit on the whole query in frontend middlewares, perhaps out of necessity since `queryExpr` wasn't exposed, but maybe because we didn't want to expose it. There are comments like https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/querysharding.go#L291 saying we need to parse it each time because we run something that modifies the input expression in-place. It's probably fine in this case because this middleware is very early in the flow and here we only examine and don't modify the input expression, but perhaps exposing it might lead to us using it in other places and risking `queryExpr` being changed unintentionally. WDYT?

Reply:
Added a comment for us to be careful when using the results of `GetQueryExpr()`.

Reply:
It should be okay to parse again here to unblock this PR, but for the long term it might be good to streamline how we pass information in the query-frontend.


> We seem to run parser.ParseExpr a fair bit on the whole query in frontend middlewares

That is true and maybe adding another parsing step may not be a dealbreaker for this one. We noticed that while working on [this](https://github.com/grafana/mimir/pull/9786) and @francoposa was looking to streamline how we pass around information between middlewares and how we handle errors.

> we run something that modifies the input expression in-place

It would be good to find all the places that modify these expr in-place and probably make them "safe", as in don't modify in place.